### ISSUE TITLE
Storing element Points in FILE I/O

### ISSUE THREAD

**Issue by roigcarlo** [simphony/simphony-common#35]
The number of points that can be stored for a given element now is set in the code

Ex: 

~~~~{.python}
MAX_POINTS_IN_FACE = 3
MAX_POINTS_IN_CELL = 4
~~~~

Points are then sotred as

~~~~{.python}
points_uuids = tables.StringCol(32, pos=2, shape=(MAX_POINTS_IN_EDGE,) )
~~~~

We agreed in the last meeting that this need to be changed.

Currently we are considering two options:

* Increase the maximum number to fit the requirements of any element.
* Use VLArray  ( variale lenght arrays ) see  https://github.com/simphony/simphony-common/issues/22#issuecomment-70848123

While I preffer to use VLArray, I can recall that someone argued that this will decrease the performance of the code.

Please let me know what you think.

**Comment by itziakos** [comment]
> While I preffer to use VLArray, I can recall that someone argued that this will decrease the performance of the code.

I also prefer using a VLArray.

The only reason that I can think of is that reading a record will need reading the file in two different points, but on the other hand hdf5 caching and the fact that records will be smaller might negate this behaviour. I would think that performance will depend on the usecase.

**Comment by nathanfranklin** [comment]
VLArray sounds good.  However, I don't totally understand it and am curious how it would be specifically used in this case.

For example, in the table (`self._group.cells`) the ids of the cells are stored.  Would there then be an associated VLArray called something like `self._groups.cell_points` which would contain the point uuids for each cell?  Did I understand that correctly?

If so, then how would the connection between these cells uuids and the rows of the VLArray be made and maintained (e.g. during deletes of cells).  Would the row index of the VLArray (`self._groups.cell_points`) be saved to the row of a particular cell in `self._group.cells`.  

And could the number of points in a particular cell change (e.g. cell1 has 5 points..then later 6 points)?  In the documentation there is "you can only update values with exactly the same size in bytes than the original row." which might be a problem then.

**Comment by itziakos** [comment]
> For example, in the table (self._group.cells) the ids of the cells are stored. Would there then be an associated VLArray called something like self._groups.cell_points which would contain the point uuids for each cell? Did I understand that correctly?

Yes the VLArray will be used to store the list of uuids in the `points` attribute of the `Cell`.

> If so, then how would the connection between these cells uuids and the rows of the VLArray be made and maintained (e.g. during deletes of cells). Would the row index of the VLArray (self._groups.cell_points) be saved to the row of a particular cell in self._group.cells.

Yes. The cells table will have the `points` column as an int value designating the VLArray row containing the points uuids.

> And could the number of points in a particular cell change (e.g. cell1 has 5 points..then later 6 points)? In the documentation there is "you can only update values with exactly the same size in bytes than the original row." which might be a problem then.

Hmm.. this is a good point, I missed that in the documentation. I did a quick check and it is true that it does not allow one to change the number of items of an already added row.

It looks that we cannot avoid the Max number of points column then.

@roigcarlo, I think that in this case do not have any option, thus for the implementation looks that the choice has been made for us.

**Comment by nathanfranklin** [comment]
If that is the case, then @khiltunen would be interested in this issue as I think he was in the discussion on what are the appropriate maximum number of points in each element

**Comment by itziakos** [comment]
Just to summarize we have to choices

**(1)** Using VLArray we do not have to set a maximum number of points, but the VLArray is treated as write-once/read-many where deleting an element does not remove the array row and updating the points of an element can potentially need to add a new row. Thus depending on the usecase the size of the VLArray will always grow unless we execute a cleanup operation (which can happen at file close or when the number of unused rows crosses a threshold).

**(2)** Using a maximum number of points column will not need to do cleanups but will heart a little the size of the table and will have to update the file format specification every time we change these values.

**Comment by khiltunen** [comment]
If we restrict to common linear element types( as i think we decided earlier) we could use MAX_POINTS_CELL=8, MAX_POINTS_FACE=4. For higher order elements the degrees of freedom can be of course more. One possible scenario where the number of points in cell could change is p-adaptive method, but that also means that we should have higher order elements in use.

**Comment by itziakos** [comment]
So what it is the consensus?

looking back and considering the implementation effort and limitations of VLArrays, I think that the maximum point options is the easiest to implement thus I now vote for option (2) in https://github.com/simphony/simphony-common/issues/35#issuecomment-71056909. So that we can get a first version of the io format early.

@roigcarlo, @nathanfranklin  and @khiltunen what is your opinion?

**Comment by roigcarlo** [comment]
If there is an implementation issue with VLArray, it seems clear that we have to go for the fixed max_number.

**Comment by khiltunen** [comment]
I agree to go with option (2)

**Comment by nathanfranklin** [comment]
Option 2 sounds good to me too

**Comment by nathanfranklin** [comment]
Closed with #36
